pytoda/smiles/smiles_language.py

```diff
@@ -190,14 +316,53 @@ def add_smi(self, smi_filepath: str, chunk_size: int = 100000) -> None:
                     self.add_smiles(smiles)
         except Exception:
             raise KeyError(
-                ".smi file needs to have 2 columns, first with IDs, second "
-                "with SMILES."
+                ".smi file needs to have 2 columns, first with SMILES, second"
+                " with IDs."
+            )
+
+    def add_dataset(self, dataset: Iterable):
+        """
+        Add a set of SMILES from an iterable.
+        The smiles_transforms are applied here in contrast to adding from .smi.
+
+        Collects and warns about invalid SMILES, and warns on finding new
+        tokens.
+
+        Args:
+            dataset (Iterable): returning SMILES strings.
+        """
+        num_tokens = len(self.token_to_index)
+
+        self.invalid_molecules = []
+        for index, smiles in enumerate(dataset):
+            smiles = self.transform_smiles(smiles)
+            self.add_smiles(smiles)
```

Comment:
I see the selfie situation is not ideal, but let's pinpoint the issue. I don't follow 100%.
I see potential issues in two places (please correct me if I under- or overcomplicate the problems):

- adding superfluous tokens because we add both smiles and selfie when only tokens for one are needed. This might happen now in the SMILESTokenizerDataset because both add_smis and add_dataset (with the translation) are called. I will remove the call to add_smis, I think it was a bad idea to put it anyway now. I vaguely remember a test failing because some token was not added with the `add_dataset` only, but it might have been a bad test (like a case where all C go to c and C is not added to the vocab but tested for).

- in the code here, `Chem.MolFromSmiles(smiles)` will report valid selfies.
I think it makes sense to separate the cases where the source string from the dataset is smiles vs selfies.

In case if SMILES, the only usecase of the selfies tokenization is in the SMILESTokenizer where we can discover it from the `selfies` attribute (I assume they must be used together). But we can put the code also in the base SMILESLanguage like this `if getattr(self, 'selfies', False):` to condition on calling the  SMILES-SELFIE translation, and even check for validity of selfies.

The case of source selfies I see only with the SELFIESLanguage right now, but please correct me if I'm wrong.
In this case we could overload the `add_dataset` of the class to check for valid selfies only.

Did I miss anything?

Reply:
One issue I see is that the `SelfiesLanguage` class should set the transforms to include selfies in the constructor. Basically `self.transform_smiles` should not be an empty `Compose` in case of selfies. Otherwise I think that `add_dataset` would add smiles tokens instead of selfies tokens to the language (unless, of course, the transforms are set explicitly which we clearly shouldn't expect from a user). 

The problem here is that we can't really separate the tokenization from the SMILESLanguage itself, that's why the whole `SMILESTokenizer` class seems somewhat artificial to me.

Reply:
> One issue I see is that the SelfiesLanguage class should set the transforms to include selfies in the constructor.
Ah ok so we always start with smiles.

Have a look now.

Reply:
hmm but we have a problem here for invalid smiles I haven't considered:
when the smiles is not valid, the smiles_transformation might fail too. The idea should be to check for invalid smiles **before** the call to the transformation.

I'm changing it so that the transforms are called in the add_smiles function, so the effect will be the same for add_smis and add_dataset.
When `selfies is True` ends up to a stacked call like this:
the selfie encoding fails.
`Chem.MolFromSmiles(selfies_decoder(selfies_encoder('C(')))` failing

Reply:
hmm but we have a problem here for invalid smiles I haven't considered:  
when the smiles is not valid, the smiles_transformation might fail too. Also in the case with selfies invalid smiles it can be detected with the call to the transformations.

Reply:
I separated the concept of invalid molecules (since we always source smiles, we simply check validity of the source smiles) and failed transformations (which could potentially have any unrelated reason) and only adding smiles which pass both hurdles.
I'm assuming if the transformations do not raise an error, they return a smiles that can be added. Or should we check for validity again here? (This would require to know if it's a selfie again, which is avoided in the latest changes at time of writing) I don't have a test case where smiles is valid but transforms fail...

>The problem here is that we can't really separate the tokenization from the SMILESLanguage itself, that's why the whole SMILESTokenizer class seems somewhat artificial to me.
The inheriting classes only define the transforms, tokenization is purely in the base. I think the changes reflect this better now.

Reply:
I mean I get that the choice of tokenizer might depend on the choice of transforms, but this is user responsibility for me.

Reply:
TODO add_smi